### PR TITLE
coll: fix ialltoallv_scattered & ialltoall_ring

### PR DESCRIPTION
**Pull request by hzhou**

### CODE REVIEW

File: src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h [hunk 1]
```diff
@@ -50,13 +50,12 @@ int MPIR_TSP_Ialltoallv_sched_intra_scattered(const void *sendbuf, const int sen
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIR_TSP_IALLTOALLV_SCHED_INTRA_SCATTERED);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIR_TSP_IALLTOALLV_SCHED_INTRA_SCATTERED);
 
-    if (bblock <= 0) {
+    /* Alltoall: each process need send #size msgs and recv #size msgs
```
Comment (hzhou) [inline src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h]: This was a previous (irrelevant) fix before I really understood the code. I am reverting to original code and adding comment to clarify the purpose.

File: src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h [hunk 2]
```diff
@@ -93,10 +92,10 @@ int MPIR_TSP_Ialltoallv_sched_intra_scattered(const void *sendbuf, const int sen
...
     for (i = bblock; i < size; i += batch_size) {
         ww = MPL_MIN(size - i, batch_size);
-        index = 0;      /* Store vtcs array from the start */
-        /* Get the send and recv ids from previous sends/recvs.
-         * ((i + j) % bblock) would ensure extracting and storing
-         * dependencies in order from '0, 1,...,bblock' */
+        index = 0;
+        /* Add dependency to ensure not to run until previous block the batch portion
+           is finsihed -- effectively limiting the on-going tasks to bblock 
+        */
```
Comment (hzhou) [inline src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h]: The original comment does not address the real purpose of the code. There is no real dependency here. Dependency is added to effectively limit the on-going jobs.

File: src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h [hunk 3]
```diff
@@ -105,11 +104,11 @@ int MPIR_TSP_Ialltoallv_sched_intra_scattered(const void *sendbuf, const int sen
         for (j = 0; j < ww; j++) {
             src = (rank + i + j) % size;
             recv_id[(i + j) % bblock] =
-                MPIR_TSP_sched_irecv((char *) recvbuf + rdispls[src],
+                MPIR_TSP_sched_irecv((char *) recvbuf + rdispls[src] * recvtype_extent,
```
Comment (hzhou) [inline src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h]: This is the real bug. I was staring at it for hours and failed to spot the obvious. This is a high example that having lines span multiple lines hinders the reading.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@raffenet Seems the only remaining test that fails is the `iallreduce ring`  on the `ch3_mxm` build; the error is an `iov` assertion. Someone else need investigate that.

File: src/mpi/coll/ialltoall/ialltoall.c [hunk 4]
```diff
@@ -285,17 +285,13 @@ int MPIR_Ialltoall_impl(const void *sendbuf, int sendcount,
         /* intracommunicator */
         switch (MPIR_Ialltoall_intra_algo_choice) {
             case MPIR_IALLTOALL_INTRA_ALGO_GENTRAN_RING:
-                mpi_errno =
+                return
```
Comment (raffenet) [inline src/mpi/coll/ialltoall/ialltoall.c]: From what I can tell, returning from the switch statement is not common across the MPICH collective code. The typical pattern I see is

> 		if (mpi_errno)
>                     MPIR_ERR_POP(mpi_errno);
> 		goto fn_exit;
> 		break;
> 

I think we shouldn't break the convention unless we plan to modify all the collectives.

File: src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h [hunk 5]
```diff
@@ -93,10 +92,10 @@ int MPIR_TSP_Ialltoallv_sched_intra_scattered(const void *sendbuf, const int sen
     /* Post more send/recv pairs as the previous ones finish */
     for (i = bblock; i < size; i += batch_size) {
         ww = MPL_MIN(size - i, batch_size);
-        index = 0;      /* Store vtcs array from the start */
-        /* Get the send and recv ids from previous sends/recvs.
-         * ((i + j) % bblock) would ensure extracting and storing
-         * dependencies in order from '0, 1,...,bblock' */
+        index = 0;
+        /* Add dependency to ensure not to run until previous block the batch portion
+         * is finsihed -- effectively limiting the on-going tasks to bblock
```
Comment (raffenet) [inline src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h]: s/finsihed/finished

**Comment by hzhou** [comment]
test:jenkins/ch3/most

File: src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h (hunk 5 above)
Comment (hzhou) [reply src/mpi/coll/ialltoallv/ialltoallv_tsp_scattered_algos.h]: thanks.

File: src/mpi/coll/ialltoall/ialltoall.c (hunk 4 above)
Comment (hzhou) [reply src/mpi/coll/ialltoall/ialltoall.c]: ok

**Comment by hzhou** [comment]
test:jenkins/ch3/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most

**Comment by pavanbalaji** [comment]
There are some tests that are still failing based on the new code.  @hzhou can you simply revert the bad commits in `mpich/master` for now, so other PRs can proceed?

**Comment by hzhou** [comment]
> There are some tests that are still failing based on the new code. @hzhou can you simply revert the bad commits in `mpich/master` for now, so other PRs can proceed?

The only one fails is the `iallreduce_tsp_ring` algorithm that only fails 1/9 allreduce tests and only on the `ch3-mxm` build. The error is an assertion error on `iov > 0`. Seems an overkill to revert it. How about an `xfail` on it?

**Comment by pavanbalaji** [comment]
xfail doesn’t make sense.  That test didn’t fail before your PR and now fails after your PR.  Ergo, your PR is broken.  Please revert it.

xfail is used very sparingly.

**Comment by pavanbalaji** [comment]
Oh, I missed your comment about merge conflicts making it a large piece of work.  That’s a different story then.  I guess we’ll just wait for you to fix it.

**Comment by hzhou** [comment]
> Oh, I missed your comment about merge conflicts making it a large piece of work. That’s a different story then. I guess we’ll just wait for you to fix it.

The truth is, I have no idea for that bug. It is only triggered `ch3-mxm` build, which I am not familiar at all. Can we do a `xfail` so other PRs can proceed while we figure out the fix (or revert)?

**Comment by hzhou** [comment]
For the record, all tests have passed other than `review-ch3-mxm` for the single `iov` assertion error. I just pushed another commit to disable that test for now so other PRs can proceed. 

@wesbland @zhenggb72 Please make a schedule to fix this bug: `allreduce ring algorithm`.  We might have to revert that code if the fix is not on the horizon.

**Comment by hzhou** [comment]
test:jenkins/ch3/mxm

**Comment by wesbland** [comment]
@tarudoodi was the author here. Apparently (on some previous PR) there was discussion about this failure. I don't know all of the history, but she wasn't able to find the issue. If you want to revert it, that's fine. if you want to xfail it, that's fine.

If it's only failing for CH3/mxm, I'd be more inclined to say the latter, but that's up to you.

**Comment by pavanbalaji** [comment]
I'd not recommend an xfail (which is easily forgotten and ignored).  I recommend reverting it.  If reverting just this patch is painful, I recommend reverting all the collective patches (i.e., the whole PR).  The PR should never have been merged in with these failures.

**Comment by wesbland** [comment]
> I'd not recommend an xfail (which is easily forgotten and ignored). I recommend reverting it. If reverting just this patch is painful, I recommend reverting all the collective patches (i.e., the whole PR). The PR should never have been merged in with these failures.

If you're having trouble reverting the offending patch, we can help with that. There's no reason to take out the unrelated algorithms.

**Comment by pavanbalaji** [comment]
> > I'd not recommend an xfail (which is easily forgotten and ignored). I recommend reverting it. If reverting just this patch is painful, I recommend reverting all the collective patches (i.e., the whole PR). The PR should never have been merged in with these failures.
> 
> If you're having trouble reverting the offending patch, we can help with that. There's no reason to take out the unrelated algorithms.

That's fine too.  Can someone do that soon?  This is holding up other development.

**Comment by tarudoodi** [comment]
We have spent some time trying to debug the failure with `ch3/mxm` test but haven't found the source of the bug yet.

**Comment by pavanbalaji** [comment]
> We have spent some time trying to debug the failure with `ch3/mxm` test but haven't found the source of the bug yet.

For future reference, I'd recommend marking such PRs that have failing tests as "do not merge".

**Comment by tarudoodi** [comment]
> > We have spent some time trying to debug the failure with `ch3/mxm` test but haven't found the source of the bug yet.
> 
> For future reference, I'd recommend marking such PRs that have failing tests as "do not merge".

I will keep that in mind for future. I didn't think that a  PR with failed tests would get merged.

**Comment by pavanbalaji** [comment]
> > For future reference, I'd recommend marking such PRs that have failing tests as "do not merge".
> 
> I will keep that in mind for future. I didn't think that a PR with failed tests would get merged.

Yes, we should have been more careful too.

**Comment by wesbland** [comment]
I think I've added the patches that you want reverted here. I left in the `test: disable test `allred7` for iallreduce ring` patch. I assume that probably needs to come out, but I'll let @hzhou do it.

test:jenkins/ch3/mxm

**Comment by pavanbalaji** [comment]
Thanks, @wesbland.  Can you remove the xfail before restarting the tests?

**Comment by hzhou** [comment]
I don't think it is really a big issue here -- especially since it all point to me ultimately :) 

I'll prepare a revert PR. It is not difficult, I was just saying it will take a day or so to pass through the review process, therefore a temporary xfail for now.

**Comment by hzhou** [comment]
> Thanks, @wesbland. Can you remove the xfail before restarting the tests?

It was not actually a `xfail`, it is an individual commit that commented out the particular test. You can simply uncomment it or remove that commit.

**Comment by wesbland** [comment]
> Thanks, @wesbland. Can you remove the xfail before restarting the tests?

@pavanbalaji There wasn't an xfail. @hzhou had commented out some of the script to generate tests for the new algorithms. I left that patch in because I wanted someone to confirm that I reverted the right patch. Either way, it should make no difference for the test results here.

If @hzhou says I reverted the right ones, he (or I) can drop that patch afterward.

**Comment by hzhou** [comment]
Oh I see, @wesbland has added the revert commit. :)

> We have spent some time trying to debug the failure with `ch3/mxm` test but haven't found the source of the bug yet.

There is nothing mentioned in the original PR #3438. For future reference, such background information is important to share. In addition, any research, motivation and analysis is also important for the code review. There is no need to put in extensive details, but there need to be a reference or mention.

**Comment by hzhou** [comment]
> We have spent some time trying to debug the failure with `ch3/mxm` test but haven't found the source of the bug yet.

There is nothing mentioned in the original PR #3438. For future reference, such background information is important to share. In addition, any research, motivation and analysis is also important for the code review. There is no need to put in extensive details, but there need to be a reference or mention.

**Comment by hzhou** [comment]
All mxm tests have passed. I am removing the extra commits, which should make no difference, and force pushed. Running tests one more time to ensure:

test:jenkins/ch3/mxm

**Comment by pavanbalaji** [comment]
@hzhou Is this ready to be merged?

**Comment by hzhou** [comment]
> @hzhou Is this ready to be merged?

Yep, the tests looks good.

**Comment by pavanbalaji** [comment]
@hzhou The comment addition in `ialltoall_tsp_ring_algos.h` is not related to the other change in that patch and should be moved to a separate patch.

**Comment by pavanbalaji** [comment]
No need to rerun the tests once you split that patch.  You can add my signoffs on all patches once you do that.

**Comment by hzhou** [comment]
merged.

**Pull request merged**